kinto_attachment/utils.py

```diff
@@ -116,32 +118,59 @@ def delete_attachment(request, link_field=None, uri=None):
     storage.delete_all("", FILE_LINKS, filters=filters, with_deleted=False)
 
 
-def save_file(content, request, randomize=True):
+def save_file(content, request, randomize=True, gzipped=False):
     folder_pattern = request.registry.settings.get('attachment.folder', '')
     folder = folder_pattern.format(**request.matchdict) or None
 
+    # Read file to compute hash.
+    content.file.seek(0)
+    filecontent = content.file.read()
+
+    if gzipped:
+        original = {
+            'filename': content.filename,
+            'hash': sha256(filecontent),
+            'mimetype': content.type,
+            'size': len(filecontent),
+        }
+        mimetype = 'application/x-gzip'
+        filename = content.filename + '.gz'
+
+        # in-memory gzipping
+        out = BytesIO()
+        with gzip.GzipFile(fileobj=out, mode="w") as f:
+            f.write(filecontent)
+
+        filecontent = out.getvalue()
+    else:
+        original = None
+        mimetype = content.type
+        filename = content.filename
+
+    # changing content so request.attachment.save() gets the gzip
+    content.file = BytesIO(filecontent)
```

Comment:
nit: You might just want to use the content and the out file descriptor rather than creating a new one here.